docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,258 @@
+walks through a nice example of creating a custom `FacialLandmarkDataset` class as
+a subclass of `Dataset`.
+
+PyTorch’s [TensorDataset](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#TensorDataset)
+is a `Dataset` wrapping tensors.
+`torch.utils.data.DataLoader` is an iterator that provides all these features.
+
+- Batching the data
+- Shuffling the data
+- Load the data in parallel using multiprocessing workers.
+
+```python
+# Mnist digits dataset
+train_data = torchvision.datasets.MNIST(
+    root='./mnist/',
+    train=True,                                     # this is training data
+    transform=torchvision.transforms.ToTensor(),    # Converts a PIL.Image or numpy.ndarray to
+                                                    # torch.FloatTensor of shape (C x H x W) and normalize in the range [0.0, 1.0]
+    download=DOWNLOAD_MNIST,
+)
+```
+
+### Training
+
+Next, we input `train_data` into neural network and get output by `forward()`, and
+finally calculate the error. The code below omits the part of calculating the accuracy.
+If you want to take a closer look at the accuracy code, please go to me see all the
+code on github.
+
+```python
+optimizer = torch.optim.Adam(cnn.parameters(), lr=LR)   # optimize all cnn parameters
+loss_func = nn.CrossEntropyLoss()   # the target label is not one-hotted
+
+# training and testing
+for epoch in range(EPOCH):
+    for step, (b_x, b_y) in enumerate(train_loader):   # 分配 batch data, normalize x when iterate train_loader
+        output = cnn(b_x)               # cnn output
+        loss = loss_func(output, b_y)   # cross entropy loss
+        optimizer.zero_grad()           # clear gradients for this training step
+        loss.backward()                 # backpropagation, compute gradients
+        optimizer.step()                # apply gradients
+```
+
+If you want the `output` of the neural network to be similar to the ground truth
+you expect, that is to keep reducing the difference between the two. This difference
+is defined by you, which is the object function or loss function. If the `loss`
+function approaches 0, then the goal is naturally achieved.
+
+## How to make ElasticDL work with PyTorch
+
+The master process of ElasticDL uses asynchronous or synchronous SGD methods to
+coordinate workers for `training`. When using asynchronous SGD method, the master
+will start a high-performance parameter server for each worker to use. When using
+synchronous SGD, ElasticDL uses Kubernetes-native's fault-tolerable AllReduce implementation.
+
+ElasticDL holds [master-worker architecture](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/overall.md#architecture).
+The master node plays the master role in two aspects.
+
+1. It's the master of the cluster.
+2. It's the master of the model training/evaluation/prediction process.
+
+### 1. Simple and Standardized Model Method
+
```

Comment:
In this section, we do not need to give too much background. We could just say:

ElasticDL requests users to provide several functions, including `forward`, `loss`, `optimizer` and `feed`. [Here](https://github.com/sql-machine-learning/elasticdl/blob/develop/model_zoo/mnist/mnist_subclass.py) is a MNIST model written in TensorFlow Keras API. 

In PyTorch, we follow the same interface design, and the following is a detailed example.